### ISSUE TITLE
Add jqwik Module for Generation of Dates and Times

### ISSUE THREAD

**Issue by jlink** [jlink/jqwik#140]
## Testing Problem

Generating dates and times with certain constraints is a standard problem when building domain-specific
generators. Currently jqwik requires users to generate dates from numbers or other primitive data types.

## Suggested Solution

Provide a new module/artefact (id "jqwik-time") that has means to generate time  and date values.
The new module will be maintained together with jqwik core and might be a default dependency when including
the aggregate artefact "jqwik".

Foundation for generation are the types in java.time. Mappers to "old" JDK classes like Date, Timestamp, Calendar etc. should be provided for convenience.

Base package for the module is `net.jqwik.api.time`. There are two classes that serve as containers for factory methods:
- `net.jqwik.api.time.Times` with method `TimeArbitrary times()`
- `net.jqwik.api.time.Dates` with method `DateArbitrary dates()`

In the long run default generation by type should be possible for all date types and time types, including `Duration` and `Period`. There should also be annotations available for the most common time and date constraints, e.g. before, after, betwee etc.

### Date Generation Options

We ignore the beginning of the Gregorian Calendar (Feb 24 1582)

- Between start/end date (borders in- or excluded)
- Min/Max Year
- Min/Max Month
- Min/Max Day of month (what's the use-case?)
- Constrain day of week
- Generate just a year
- Generate just a month
- Generate just a day of week
- Generate a month in a year

#### Optional

- With/without lap year/lap day
- Generate Period (date-based time span)
- Map to Date
- Map to Calendar

### Time Generation Options

- Min/Max Hours
- Min/Max Minutes
- Min/Max Seconds
- Min/Max Millis
- Between given times (borders in- or excluded)
- Constrain precision (nanos, millis, seconds, minutes)
- Include/exclude lap seconds

### Instant, LocalDateTime, OffsetDateTime, ZonedDateTime

Combine date and time constraints with (optional) offset/timezone/dst generation.
Handle impossible times, e.g. between 2 and 3 am on change to DST.

## Annotations

To be discussed

## Discussion

What about non Gregorian dates?

**Comment by jlink** [comment]
dsfgsdfg

**Comment by zinki97** [comment]
Which class should be used for the date? I think the Date class is deprecated. Maybe Calendar?

Calendar and Date also have time attributes. How should we handle this?

**Comment by zinki97** [comment]
My first design of DateArbitrary:
```
public interface DateArbitrary extends Arbitrary<LocalDate> {

	default DateArbitrary between(LocalDate dateBegin, LocalDate dateEnd){
		return greaterOrEqual(dateBegin).lessOrEqual(dateEnd);
	}
	DateArbitrary greaterOrEqual(LocalDate date);
	DateArbitrary lessOrEqual(LocalDate date);

	DateArbitrary yearGreaterOrEqual(int min);
	DateArbitrary yearLessOrEqual(int max);

	DateArbitrary monthGreaterOrEqual(int min);
	DateArbitrary monthLessOrEqual(int max);

	DateArbitrary dayOfMonthGreaterOrEqual(int min);
	DateArbitrary dayOfMonthLessOrEqual(int max);

	DateArbitrary onlyDaysOfWeek(DayOfWeek... dayOfWeek);

	DateArbitrary onlyGenerateYears();
	DateArbitrary onlyGenerateMonths();
	DateArbitrary onlyGenerateDaysOfMonth();
	DateArbitrary onlyGenerateYearsAndMonths();

}
```

**Comment by jlink** [comment]
I think LocalDate is the best choice

**Comment by jlink** [comment]
What about atTheEarliest/atTheLatest rather than greaterOrEqual/lessOrEqual.
Maybe before/after is enough since there‘s the between for the cases including the borders.

**Comment by jlink** [comment]
The only* methods should probably be static methods on the `Dates` class.

**Comment by zinki97** [comment]
> The only* methods should probably be static methods on the `Dates` class.

How should they look like?

**Comment by jlink** [comment]
> > The only* methods should probably be static methods on the `Dates` class.
> 
> How should they look like?

I’d probably go with years(), daysOfWeek(), months() etc.

**Comment by jlink** [comment]
Added new module "jqwik-time" in https://github.com/jlink/jqwik/commit/b2b68ce055761a9f0d8fc5559753666f7f884c49

**Comment by zinki97** [comment]
Working on it https://github.com/jlink/jqwik/pull/141

**Comment by zinki97** [comment]
### Date Generation Options

* [x] Between start/end date
* [x] Min/Max Year
* [x] Min/Max Month
* [x] Min/Max Day of month
* [x] Constrain day of week
* [x] Generate just a year
* [x] Generate just a month
* [x] Generate just a day of month
* [x] Generate just a day of week
* [x] Generate a month in a year
* [x] Generate a day in a 

 
#### Optional
* [ ]  With/without lap year/lap day
* [ ]  Generate Period (date-based time span)
* [ ]  Map to Date
* [ ]  Map to Calendar

### Time Generation Options
* [ ]  Min/Max Hours
* [ ]  Min/Max Minutes
* [ ]  Min/Max Seconds
* [ ]  Min/Max Millis
* [ ]  Between given times (borders in- or excluded)
* [ ]  Constrain precision (nanos, millis, seconds, minutes)
* [ ]  Include/exclude lap seconds

**Comment by jlink** [comment]
Version 1.4.0-SNAPSHOT is available with the latest dates generation features

**Comment by zinki97** [comment]
> Map to Date

I think the Date class is deprecated. Should we add it anyway?

**Comment by jlink** [comment]
Yes, it's so widely in use.

Am Di., 12. Jan. 2021 um 15:58 Uhr schrieb zinki97 <notifications@github.com

**Comment by zinki97** [comment]
> Generate Period (date-based time span)

I do not understand this point. Can you give me an example, please?

**Comment by jlink** [comment]
https://docs.oracle.com/javase/8/docs/api/java/time/Period.html

Von meinem iPad gesendet

**Comment by zinki97** [comment]
We have decided to create a new issue.

--- closed --- (I can't close this issue)

**Comment by jlink** [comment]
Date generation has been completely implemented in 1.4.0-SNAPSHOT.

**Comment by jlink** [comment]
Follow up for time generation in https://github.com/jlink/jqwik/issues/154
